frontend/src/metabase/query_builder/actions/core/updateQuestion.ts

```diff
@@ -29,15 +29,22 @@ import { onCloseQuestionInfo, setQueryBuilderMode } from "../ui";
 
 import { getQuestionWithDefaultVisualizationSettings } from "./utils";
 
-function hasNewColumns(question: Question, queryResult: Dataset) {
-  // NOTE: this assume column names will change
+function hasNewColumns(question: Question, queryResult: Dataset | null) {
+  if (!queryResult) {
+    return false;
+  }
+
+  const queryResultColumns = queryResult.data.cols;
+  if (question.isNative()) {
+    return queryResultColumns.length > 0;
+  }
+
+  // NOTE: this assumes column names will change
   // technically this is wrong because you could add and remove two columns with the same name
-  const query = question.legacyQuery({ useStructuredQuery: true });
-  const previousColumns =
-    (queryResult && queryResult.data.cols.map(col => col.name)) || [];
-  const nextColumns =
-    query instanceof StructuredQuery ? query.columnNames() : [];
-  return _.difference(nextColumns, previousColumns).length > 0;
+  const query = question.query();
+  const stageIndex = -1;
+  const queryColumns = Lib.returnedColumns(query, stageIndex);
+  return queryColumns.length > queryResultColumns.length;
```

Comment:
I think this is not enough. Previously diff of arrays of names of columns was checked.
So it does not handle a case where 1 column is removed and another is added.

Reply:
nice catch btw

Reply:
I'm experimenting to see if I can avoid the issue with the wrong column names in 
`returnedColumns`. But it seems that not. Reverted this change. The PR would have to stay here until the referenced MLv2 bug is fixed